In [1]:
#imports
import pandas as pd

In [2]:
circuits_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/circuits.csv')

constructor_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/constructors.csv')
constructor_results_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/constructor_results.csv')
constructor_standings_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/constructor_standings.csv')

drivers_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/drivers.csv')
drivers_standings_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/driver_standings.csv')

lap_times_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/lap_times.csv')

qualifying_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/qualifying.csv')

races_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/races.csv')
result_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/results.csv')

season_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/seasons.csv')

sprint_results_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/sprint_results.csv')

status_df = pd.read_csv('../input/formula-1-world-championship-1950-2020/status.csv')

## Lewis Hamilton Records
(Principios 2025 antes de comenzar la temporada del 2025)
(Early 2025 before the start of the 2025 season)

### El piloto con más puntos actualmente || The driver currently with the most points

In [3]:
drivers_standings_df['year'] = drivers_standings_df['raceId'].map(races_df.set_index('raceId')['year'])

# Tomar solo el último registro de cada temporada para cada piloto (el acumulado final de la temporada)
# Take only the last record of each season for each driver (the final accumulated record of the season).
last_standing_per_season = drivers_standings_df.sort_values(by=['year', 'raceId']).drop_duplicates(subset=['year', 'driverId'], keep='last')

# Sumar los puntos por piloto
# Sum up the points per driver
total_points = last_standing_per_season.groupby('driverId')['points'].sum().reset_index()

# Unir con la tabla de pilotos para obtener nombres
# # Join with drivers table to get names
drivers_df = drivers_df[['driverId', 'forename', 'surname']]
total_points = total_points.merge(drivers_df, on='driverId')

# Ordenar de mayor a menor
# Sort from highest to lowest
top_driver = total_points.sort_values(by='points', ascending=False).iloc[0]

print(f"🏁 El piloto con más puntos es {top_driver['forename']} {top_driver['surname']} con un total de {top_driver['points']} puntos. 🏁")
print(f"🏁 The driver with the most points is {top_driver['forename']} {top_driver['surname']} with a total of {top_driver['points']} points. 🏁")

🏁 El piloto con más puntos es Lewis Hamilton con un total de 4862.5 puntos. 🏁
🏁 The driver with the most points is Lewis Hamilton with a total of 4862.5 points. 🏁


### El piloto con más carreras ganadas || The driver with the most race wins

In [4]:
# Filtrar solo las victorias
# # Filter only wins
wins_df = result_df[result_df['positionOrder'] == 1]

# Contar las victorias por piloto
# Count wins per driver
wins_count = wins_df['driverId'].value_counts().reset_index()
wins_count.columns = ['driverId', 'wins']

# Unir con la tabla de pilotos para obtener nombres
# Join with drivers table to get names.
drivers_df = drivers_df[['driverId', 'forename', 'surname']]
wins_count = wins_count.merge(drivers_df, on='driverId')

# Ordenar por número de victorias
# Sort by number of wins
top_winner = wins_count.sort_values(by='wins', ascending=False).iloc[0]

print(f"🚀 El piloto con más victorias es {top_winner['forename']} {top_winner['surname']} con {top_winner['wins']} carreras ganadas. 🚀")
print(f"🚀 The driver with the most wins is {top_winner['forename']} {top_winner['surname']} with {top_winner['wins']} races won. 🚀")

🚀 El piloto con más victorias es Lewis Hamilton con 105 carreras ganadas. 🚀
🚀 The driver with the most wins is Lewis Hamilton with 105 races won. 🚀


### El piloto con más podios || The driver with the most podiums

In [5]:
# Filtrar solo las posiciones de podio (1.º, 2.º o 3.º lugar)
# Filter only podium positions (1st, 2nd or 3rd place)
podiums_df = result_df[result_df['positionOrder'].isin([1, 2, 3])]

# Contar la cantidad de podios por piloto
# Count the number of podiums per driver.
podium_count = podiums_df['driverId'].value_counts().reset_index()
podium_count.columns = ['driverId', 'podiums']

# Unir con la tabla de pilotos para obtener nombres
# Join with drivers table to get names.
drivers_df = drivers_df[['driverId', 'forename', 'surname']]
podium_count = podium_count.merge(drivers_df, on='driverId')

# Ordenar por número de podios
# Sort by number of podiums
top_podium_driver = podium_count.sort_values(by='podiums', ascending=False).iloc[0]

print(f"🔥 El piloto con más podios es {top_podium_driver['forename']} {top_podium_driver['surname']} con {top_podium_driver['podiums']} podios.🔥")
print(f"🔥 The driver with the most podiums is {top_podium_driver['forename']} {top_podium_driver['surname']} with {top_podium_driver['podiums']} podiums.🔥")


🔥 El piloto con más podios es Lewis Hamilton con 202 podios.🔥
🔥 The driver with the most podiums is Lewis Hamilton with 202 podiums.🔥


### El piloto con más pole positions || The driver with the most pole positions
Pole positions = Salir primero en una carrera || Starting first in a race

In [6]:
pole_positions_df = qualifying_df[qualifying_df['position'] == 1]

# Contar cuántas pole positions tiene cada piloto
# Count how many pole positions each driver has
pole_count = pole_positions_df['driverId'].value_counts().reset_index()
pole_count.columns = ['driverId', 'pole_positions']

# Unir con la tabla de pilotos para obtener nombres
# Join with drivers table to get names.
drivers_df = drivers_df[['driverId', 'forename', 'surname']]
pole_count = pole_count.merge(drivers_df, on='driverId')

# Ordenar por número de poles
# Sort by number of poles
top_pole_driver = pole_count.sort_values(by='pole_positions', ascending=False).iloc[0]

print(f"🏆 El piloto con más pole positions es {top_pole_driver['forename']} {top_pole_driver['surname']} con {top_pole_driver['pole_positions']} pole positions. 🏆")
print(f"🏆 The driver with the most pole positions is {top_pole_driver['forename']} {top_pole_driver['surname']} with {top_pole_driver['pole_positions']} pole positions. 🏆")

🏆 El piloto con más pole positions es Lewis Hamilton con 107 pole positions. 🏆
🏆 The driver with the most pole positions is Lewis Hamilton with 107 pole positions. 🏆


### El piloto con más carreras lideradas || The driver with the most races led
(al menos una vuelta liderada en carrera || at least one lap led in a race)

In [7]:
# Filtrar vueltas donde un piloto estuvo en P1
# Filter laps where a driver was in P1
laps_led = lap_times_df[lap_times_df['position'] == 1]

# Contar en cuántas carreras diferentes lideró cada piloto (cada raceId único cuenta una vez)
# Count how many different races each driver led (each unique raceId counts once)
races_led_count = laps_led.groupby('driverId')['raceId'].nunique().reset_index()
races_led_count.columns = ['driverId', 'races_led']

# Unir con la tabla de pilotos para obtener nombres
# Join with drivers table to get names.
races_led_count = races_led_count.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')

# Ordenar de mayor a menor
# Sort from highest to lowest
races_led_count = races_led_count.sort_values(by='races_led', ascending=False)

top_driver = races_led_count.iloc[0]
print(f"🔝 El piloto con más carreras lideradas es {top_driver['forename']} {top_driver['surname']} con {top_driver['races_led']} carreras lideradas. 🔝")
print(f"🔝 The driver with the most races led is {top_driver['forename']} {top_driver['surname']} with {top_driver['races_led']} races led. 🔝")

🔝 El piloto con más carreras lideradas es Lewis Hamilton con 190 carreras lideradas. 🔝
🔝 The driver with the most races led is Lewis Hamilton with 190 races led. 🔝


### El piloto con más vueltas lideradas || The driver with the most laps led
Total de vueltas en P1 || Total laps in P1

In [8]:
# Contar cuántas vueltas lideró cada piloto
# Count how many laps each driver led.
laps_led_count = laps_led['driverId'].value_counts().reset_index()
laps_led_count.columns = ['driverId', 'laps_led']

# Unir con la tabla de pilotos
# Join with driver table
laps_led_count = laps_led_count.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')

# Ordenar de mayor a menor
# Sort from highest to lowest
laps_led_count = laps_led_count.sort_values(by='laps_led', ascending=False)

top_driver = laps_led_count.iloc[0]
print(f"🎯 El piloto con más vueltas lideradas es {top_driver['forename']} {top_driver['surname']} con {top_driver['laps_led']} vueltas lideradas. 🎯")
print(f"🎯 The driver with the most laps led is {top_driver['forename']} {top_driver['surname']} with {top_driver['laps_led']} laps led. 🎯")

🎯 El piloto con más vueltas lideradas es Lewis Hamilton con 5487 vueltas lideradas. 🎯
🎯 The driver with the most laps led is Lewis Hamilton with 5487 laps led. 🎯


### El piloto con más campeonatos mundiales || The driver with the most world championships

In [9]:
# Obtener la última carrera de cada temporada
# Get the last race of each season
last_race_per_season = races_df.groupby('year')['raceId'].max().reset_index()

# Filtrar la tabla de standings para obtener SOLO los campeones de cada temporada
# Filter the standings table to get ONLY the champions of each season
champions = drivers_standings_df[
    (drivers_standings_df['position'] == 1) & 
    (drivers_standings_df['raceId'].isin(last_race_per_season['raceId']))
]

# Contar cuántos títulos tiene cada piloto
# Count how many titles each driver has
championships_count = champions['driverId'].value_counts().reset_index()
championships_count.columns = ['driverId', 'titles']

# Unir con la tabla de pilotos para obtener nombres
# Join with drivers table to get names
championships_count = championships_count.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')

# Ordenar de mayor a menor
# Sort from highest to lowest
championships_count = championships_count.sort_values(by='titles', ascending=False)

# Obtener el máximo número de títulos
# Get maximum number of titles
max_titles = championships_count['titles'].max()

# Filtrar todos los pilotos que tengan ese número de títulos
# Filter out all drivers with that number of titles
top_drivers = championships_count[championships_count['titles'] == max_titles]

for _, row in top_drivers.iterrows():
    print(f"🏁 {row['forename']} {row['surname']} tiene {row['titles']} títulos. 🏁")
    print(f"🏁 {row['forename']} {row['surname']} has {row['titles']} titles. 🏁")


🏁 Lewis Hamilton tiene 7 títulos. 🏁
🏁 Lewis Hamilton has 7 titles. 🏁
🏁 Michael Schumacher tiene 7 títulos. 🏁
🏁 Michael Schumacher has 7 titles. 🏁


### El piloto con mas victorias en diferentes temporadas || The driver with the most victories in different seasons

In [10]:
# Filtrar las carreras ganadas (positionOrder == 1 indica el ganador de la carrera)
# Filter the races won (positionOrder == 1 indicates the winner of the race)
winners = result_df[result_df['positionOrder'] == 1]

# Unir con la tabla de carreras para obtener el año de cada carrera ganada
# Merge with the table of races to get the year of each won race
winners = winners.merge(races_df[['raceId', 'year']], on='raceId')

# Contar en cuántas temporadas diferentes ganó al menos una carrera cada piloto
# Count how many different seasons each driver won at least one race in
seasons_won = winners.groupby('driverId')['year'].nunique().reset_index()
seasons_won.columns = ['driverId', 'seasons_won']

# Unir con la tabla de pilotos para obtener nombres
# Join with driver table to get names.
seasons_won = seasons_won.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')

# Ordenar de mayor a menor
# Sort from highest to lowest
seasons_won = seasons_won.sort_values(by='seasons_won', ascending=False)

# Obtener el número máximo de temporadas ganadas
# Get the maximum number of seasons won
max_seasons = seasons_won['seasons_won'].max()

# Filtrar todos los pilotos que tengan ese número de temporadas ganadas
# Filter out all drivers with that number of seasons won
top_drivers = seasons_won[seasons_won['seasons_won'] == max_seasons]

for _, row in top_drivers.iterrows():
    print(f"🪄 {row['forename']} {row['surname']} ha ganado en {row['seasons_won']} temporadas diferentes. 🪄")
    print(f"🪄 {row['forename']} {row['surname']} has won in {row['seasons_won']} different seasons. 🪄")


🪄 Lewis Hamilton ha ganado en 16 temporadas diferentes. 🪄
🪄 Lewis Hamilton has won in 16 different seasons. 🪄


### El piloto con más victorias en x circuito || The driver with the most wins on x circuit 

In [11]:
# Filtrar las carreras ganadas (positionOrder == 1 indica el ganador de la carrera)
# Filter the races won (positionOrder == 1 indicates the winner of the race)
winners = result_df[result_df['positionOrder'] == 1]

# Unir con la tabla de carreras para obtener el circuito de cada carrera ganada
# Merge with the table of races to get the track of each won race
winners = winners.merge(races_df[['raceId', 'circuitId']], on='raceId')

# Contar cuántas veces cada piloto ha ganado en cada circuito
# Count how many times each driver has won at each circuit
circuit_wins = winners.groupby(['driverId', 'circuitId']).size().reset_index(name='wins')

# Encontrar el máximo número de victorias en un solo circuito
# Find the maximum number of wins in a single circuit
max_wins = circuit_wins['wins'].max()

# Filtrar todos los pilotos que tengan ese número de victorias en algún circuito
# Filter out all drivers who have that number of wins on any circuit
top_circuit_winners = circuit_wins[circuit_wins['wins'] == max_wins]

# Unir con la tabla de pilotos para obtener nombres
# Join with drivers table to get names
top_circuit_winners = top_circuit_winners.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')

# Unir con la tabla de circuitos para obtener el nombre del circuito
# Merge with the circuit table to get the name of the circuit
top_circuit_winners = top_circuit_winners.merge(circuits_df[['circuitId', 'name']], on='circuitId')

for _, row in top_circuit_winners.iterrows():
    print(f"🥇 {row['forename']} {row['surname']} ha ganado {row['wins']} veces en {row['name']}. 🥇")
    print(f"🥇 {row['forename']} {row['surname']} has won {row['wins']} times in {row['name']}. 🥇")


🥇 Lewis Hamilton ha ganado 9 veces en Silverstone Circuit. 🥇
🥇 Lewis Hamilton has won 9 times in Silverstone Circuit. 🥇


### El piloto con más victorias como rookie || The driver with the most wins as a rookie

In [12]:
# Obtener la primera temporada de cada piloto (año de su primera carrera)
# Get the first season of each driver (year of his first race)
first_season = races_df[['raceId', 'year']].merge(result_df[['raceId', 'driverId']], on='raceId')
first_season = first_season.groupby('driverId')['year'].min().reset_index()
first_season.columns = ['driverId', 'rookie_year']

# Filtrar las victorias (positionOrder == 1) y unir con la temporada de debut
# Filter wins (positionOrder == 1) and join with debut season.
rookie_wins = result_df[result_df['positionOrder'] == 1].merge(races_df[['raceId', 'year']], on='raceId')
rookie_wins = rookie_wins.merge(first_season, on='driverId')

# Filtrar solo las victorias en la temporada de debut
# Filter out only the wins in the debut season
rookie_wins = rookie_wins[rookie_wins['year'] == rookie_wins['rookie_year']]

# Contar las victorias de cada piloto en su año de debut
# Count the wins of each driver in his debut year.
rookie_win_count = rookie_wins['driverId'].value_counts().reset_index()
rookie_win_count.columns = ['driverId', 'rookie_wins']

# Unir con la tabla de pilotos para obtener nombres
# Join with driver table to get names.
rookie_win_count = rookie_win_count.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')

# Ordenar de mayor a menor
# Sort from highest to lowest
rookie_win_count = rookie_win_count.sort_values(by='rookie_wins', ascending=False)

# Obtener el número máximo de victorias de un rookie
# Get the maximum number of wins of a rookie
max_rookie_wins = rookie_win_count['rookie_wins'].max()

# Filtrar todos los pilotos que lograron esa cantidad de victorias
# Filter out all drivers who achieved that number of wins
top_rookies = rookie_win_count[rookie_win_count['rookie_wins'] == max_rookie_wins]

for _, row in top_rookies.iterrows():
    print(f"🎖️ {row['forename']} {row['surname']} ganó {row['rookie_wins']} carreras en su temporada de debut ({row['driverId']}). 🎖️")
    print(f"🎖️ {row['forename']} {row['surname']} won {row['rookie_wins']} races in its debut season ({row['driverId']}). 🎖️")


🎖️ Lewis Hamilton ganó 4 carreras en su temporada de debut (1). 🎖️
🎖️ Lewis Hamilton won 4 races in its debut season (1). 🎖️
🎖️ Jacques Villeneuve ganó 4 carreras en su temporada de debut (35). 🎖️
🎖️ Jacques Villeneuve won 4 races in its debut season (35). 🎖️


### El piloto con más pole position como rookie || The driver with the most pole positions as a rookie

In [13]:
# Obtener la primera temporada de cada piloto (año de su primera carrera)
# Get the first season of each driver (year of his first race)
first_season = races_df[['raceId', 'year']].merge(result_df[['raceId', 'driverId']], on='raceId')
first_season = first_season.groupby('driverId')['year'].min().reset_index()
first_season.columns = ['driverId', 'rookie_year']

# Filtrar las poles (posición 1 en clasificación)
# Filter poles (position 1 in qualifying)
rookie_poles = qualifying_df[qualifying_df['position'] == 1]

# Unir con la tabla de carreras para obtener el año de la pole
# Join with the race table to get the pole year
rookie_poles = rookie_poles.merge(races_df[['raceId', 'year']], on='raceId')

# Unir con la primera temporada de cada piloto
# Merge with each driver's first season
rookie_poles = rookie_poles.merge(first_season, on='driverId')

# Filtrar solo las poles en la temporada de debut
# Filter only the poles in the debut season
rookie_poles = rookie_poles[rookie_poles['year'] == rookie_poles['rookie_year']]

# Contar las poles de cada piloto en su año de debut
# Count the poles of each driver in his debut year
rookie_pole_count = rookie_poles['driverId'].value_counts().reset_index()
rookie_pole_count.columns = ['driverId', 'rookie_poles']

# Unir con la tabla de pilotos para obtener nombres
# Join with driver table to get names
rookie_pole_count = rookie_pole_count.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')

# Ordenar de mayor a menor
# Sort from highest to lowest
rookie_pole_count = rookie_pole_count.sort_values(by='rookie_poles', ascending=False)

# Obtener el número máximo de poles de un rookie
# Get the maximum number of poles of a rookie
max_rookie_poles = rookie_pole_count['rookie_poles'].max()

# Filtrar todos los pilotos que lograron esa cantidad de poles
# Filter all drivers who achieved that number of poles
top_rookies = rookie_pole_count[rookie_pole_count['rookie_poles'] == max_rookie_poles]

# Mostrar los rookies con más pole positions
for _, row in top_rookies.iterrows():
    print(f"🚀 {row['forename']} {row['surname']} logró {row['rookie_poles']} pole positions en su temporada de debut. 🚀")
    print(f"🚀 {row['forename']} {row['surname']} achieved {row['rookie_poles']} pole positions in his debut season. 🚀")


🚀 Lewis Hamilton logró 5 pole positions en su temporada de debut. 🚀
🚀 Lewis Hamilton achieved 5 pole positions in his debut season. 🚀


### El piloto con más premios con un mismo constructor || The driver with the most awards with the same constructor

In [14]:
# Crear una nueva columna con el nombre completo del piloto
# Create a new column with the driver's full name
drivers_df['fullName'] = drivers_df['forename'] + ' ' + drivers_df['surname']

# Agrupar por piloto y constructor en result_df y contar las carreras
# Group by driver and constructor in result_df and count races.
driver_constructor_grands_prix = result_df.groupby(['driverId', 'constructorId'])['raceId'].nunique().reset_index()

# Unir con la tabla de drivers para obtener el nombre completo del piloto
# Join with the driver table to get the full driver name
driver_constructor_grands_prix = driver_constructor_grands_prix.merge(drivers_df[['driverId', 'fullName']], on='driverId', how='left')

# Unir con la tabla de constructores para obtener el nombre del constructor
# Merge with the constructor table to get the name of the constructor
driver_constructor_grands_prix = driver_constructor_grands_prix.merge(constructor_df[['constructorId', 'name']], on='constructorId', how='left')

# Encontramos el piloto con más Grandes Premios con un mismo constructor
# We find the driver with the most Grand Prix with the same constructor
max_grands_prix_constructor = driver_constructor_grands_prix.loc[driver_constructor_grands_prix['raceId'].idxmax()]

print(f"🏎️ Más Grandes Premios con un mismo constructor: {max_grands_prix_constructor['fullName']} con {max_grands_prix_constructor['raceId']} carreras en {max_grands_prix_constructor['name']}. 🏎️")
print(f"🏎️ More Grand Prix with same constructor: {max_grands_prix_constructor['fullName']} with {max_grands_prix_constructor['raceId']} races in {max_grands_prix_constructor['name']}. 🏎️")

🏎️ Más Grandes Premios con un mismo constructor: Lewis Hamilton con 246 carreras en Mercedes. 🏎️
🏎️ More Grand Prix with same constructor: Lewis Hamilton with 246 races in Mercedes. 🏎️


### El piloto con más carreras consecutivas finalizadas con puntos || The driver with the most consecutive race finishes in points

In [15]:
# Filtrar carreras donde el piloto obtuvo puntos (por ejemplo, posición 1-10)
# Filter races where the driver scored points (e.g. position 1-10)
result_with_points = result_df[result_df['positionOrder'] <= 10]

# Ordenar por piloto y raceId para asegurar que las carreras están en orden cronológico
# Sort by driver and raceId to ensure races are in chronological order.
result_with_points = result_with_points.sort_values(by=['driverId', 'raceId'])

# Crear una columna de "carrera consecutiva" para cada piloto
# Create a “consecutive race” column for each driver
result_with_points['consecutive'] = result_with_points.groupby('driverId')['raceId'].diff().ne(1).cumsum()

# Contamos las carreras consecutivas con puntos para cada piloto
# We count consecutive races with points for each driver
consecutive_points = result_with_points.groupby(['driverId', 'consecutive']).size().reset_index(name='consecutive_points')

# Encontrar el piloto con más carreras consecutivas con puntos
# Find the driver with the most consecutive races with points.
max_consecutive_points = consecutive_points.loc[consecutive_points['consecutive_points'].idxmax()]

# Obtener el nombre del piloto
# Get driver name
driver_name = drivers_df.loc[drivers_df['driverId'] == max_consecutive_points['driverId'], 'fullName'].values[0]

print(f"🏁 El piloto con más carreras consecutivas finalizadas con puntos es {driver_name} con {max_consecutive_points['consecutive_points']} carreras consecutivas. 🏁")
print(f"🏁 The driver with the most consecutive races finished with points is {driver_name} with {max_consecutive_points['consecutive_points']} consecutive races. 🏁")

🏁 El piloto con más carreras consecutivas finalizadas con puntos es Lewis Hamilton con 48 carreras consecutivas. 🏁
🏁 The driver with the most consecutive races finished with points is Lewis Hamilton with 48 consecutive races. 🏁


### El piloto con más victorias en diferentes circuitos || The driver with the most victories in different circuits 

In [16]:
# Filtrar las victorias (posición 1)
# Filter victories (position 1)
victories = result_df[result_df['positionOrder'] == 1]

# Unir con races_df para obtener el circuito (circuitId) de cada carrera
# Merge with races_df to get the circuit (circuitId) of each race
victories = victories.merge(races_df[['raceId', 'circuitId']], on='raceId', how='left')

# Agrupar por piloto y circuito para contar las victorias en diferentes circuitos
# Group by driver and circuit to count the victories in different circuits
victories_by_circuit = victories.groupby(['driverId', 'circuitId']).size().reset_index(name='victories')

# Contar las victorias por piloto en circuitos diferentes (sin contar repeticiones en el mismo circuito)
# Count victories per driver in different circuits (without counting repetitions in the same circuit)
driver_victories_in_circuits = victories_by_circuit.groupby('driverId')['circuitId'].nunique().reset_index(name='unique_victories')

# Encontrar el piloto con más victorias en circuitos diferentes
# Find the driver with the most victories in different circuits
max_victories = driver_victories_in_circuits.loc[driver_victories_in_circuits['unique_victories'].idxmax()]

# Obtener el nombre del piloto
# Get driver name
driver_name = drivers_df.loc[drivers_df['driverId'] == max_victories['driverId'], 'forename'].values[0] + ' ' + drivers_df.loc[drivers_df['driverId'] == max_victories['driverId'], 'surname'].values[0]

print(f"🚀 El piloto con más victorias en diferentes circuitos es {driver_name} con {max_victories['unique_victories']} victorias. 🚀")
print(f"🚀 The driver with the most wins on different circuits is {driver_name} with {max_victories['unique_victories']} wins. 🚀")

🚀 El piloto con más victorias en diferentes circuitos es Lewis Hamilton con 31 victorias. 🚀
🚀 The driver with the most wins on different circuits is Lewis Hamilton with 31 wins. 🚀


### El piloto con más podios en diferentes circuitos || The driver with the most podiums in different circuits

In [17]:
# Filtrar los resultados de los podios (posiciones 1, 2, 3)
# Filter the results of the podiums (positions 1, 2, 3)
podiums = result_df[result_df['positionOrder'].isin([1, 2, 3])]

# Unir con races_df para obtener el circuito (circuitId) de cada carrera
# Join with races_df to get the circuit (circuitId) of each race.
podiums = podiums.merge(races_df[['raceId', 'circuitId']], on='raceId', how='left')

# Agrupar por piloto y circuito para contar los podios en diferentes circuitos
# Group by driver and circuit to count podiums in different circuits
podiums_by_circuit = podiums.groupby(['driverId', 'circuitId']).size().reset_index(name='podiums')

# Contar los podios por piloto en circuitos diferentes (sin contar repeticiones en el mismo circuito)
# Count podiums per driver in different circuits (without counting repetitions in the same circuit)
driver_podiums_in_circuits = podiums_by_circuit.groupby('driverId')['circuitId'].nunique().reset_index(name='unique_podiums')

# Encontrar el piloto con más podios en circuitos diferentes
# Find the driver with the most podiums in different circuits
max_podiums = driver_podiums_in_circuits.loc[driver_podiums_in_circuits['unique_podiums'].idxmax()]

# Obtener el nombre del piloto
# Get driver name
driver_name = drivers_df.loc[drivers_df['driverId'] == max_podiums['driverId'], 'forename'].values[0] + ' ' + drivers_df.loc[drivers_df['driverId'] == max_podiums['driverId'], 'surname'].values[0]

print(f"💥 El piloto con más podios en diferentes circuitos es {driver_name} con {max_podiums['unique_podiums']} podios. 💥")
print(f"💥 The driver with the most podiums in different circuits is {driver_name} with {max_podiums['unique_podiums']} podiums. 💥")

💥 El piloto con más podios en diferentes circuitos es Lewis Hamilton con 36 podios. 💥
💥 The driver with the most podiums in different circuits is Lewis Hamilton with 36 podiums. 💥


## Resumen
Hamilton es el piloto actual con más records. Tras 14 años consecutivos con Mercedes, abre una nueva era con Ferrari. ¿Podrá seguir batiendo records?
Lewis Hamilton es el piloto con más record en el mundo de la formula 1. A día de hoy es considerado uno de los mejores pilotos. Comenzo en la formula 1 con McLaren en 2007 con 22 años. En 2013 firmo un contrato con Mercedes y estarían juntos hasta 2024 que decidio cumplir su mayor sueño. Poder correr con Ferrari.
A lo largo de su carrera cumplió todos los records que acabamos de ver. Seguramente tenga más record, pero necesitaríamos más información.

1. 🏁 El piloto con más puntos es Lewis Hamilton con un total de 4862.5 puntos. 🏁
2. 🚀 El piloto con más victorias es Lewis Hamilton con 105 carreras ganadas. 🚀
3. 🔥 El piloto con más podios es Lewis Hamilton con 202 podios. 🔥
4. 🏆 El piloto con más pole positions es Lewis Hamilton con 107 pole positions. 🏆
6. 🔝 El piloto con más carreras lideradas es Lewis Hamilton con 190 carreras lideradas. 🔝
7. 🎯 El piloto con más vueltas lideradas es Lewis Hamilton con 5487 vueltas lideradas. 🎯
8. 🏁 Lewis Hamilton y Michael Schumacher tienen ambos 7 títulos mundiales. ¿Será capaz Hamilton de batir el record en su nueva etapa con Ferrari? 🏁
9. 🪄 Lewis Hamilton ha ganado en 16 temporadas diferentes. 🪄
10. 🥇 Lewis Hamilton ha ganado 9 veces en Silverstone Circuit. 🥇
11. 🎖️ Lewis Hamilton y Jacques Villeneuve comparten el record de ganar 4 carreras en su temporada de debut 🎖️
12. 🚀 Lewis Hamilton logró 5 pole positions en su temporada de debut. 🚀
13. 🏎️ Más Grandes Premios con un mismo constructor: Lewis Hamilton con 246 carreras en Mercedes. 🏎️
14. 🏁 El piloto con más carreras consecutivas finalizadas con puntos es Lewis Hamilton con 48 carreras consecutivas. 🏁
15. 🚀 El piloto con más victorias en diferentes circuitos es Lewis Hamilton con 31 victorias. 🚀
16. 💥 El piloto con más podios en diferentes circuitos es Lewis Hamilton con 36 podios. 💥


## Summary
Hamilton is the current driver with the most records. After 14 consecutive years with Mercedes, he opens a new era with Ferrari. Can he continue to break records?
Lewis Hamilton is the driver with the most records in the world of Formula 1. Today he is considered one of the best drivers. He started in Formula 1 with McLaren in 2007 at the age of 22. In 2013 he signed a contract with Mercedes and they would be together until 2024 when he decided to fulfill his biggest dream. To be able to race with Ferrari.
Throughout his career he fulfilled all the records we have just seen. Surely he has more records, but we would need more information.

1. 🏁 The driver with the most points is Lewis Hamilton with a total of 4862.5 points. 🏁
2. 🚀 The driver with the most wins is Lewis Hamilton with 105 race wins. 🚀
3. 🔥 The driver with the most podiums is Lewis Hamilton with 202 podiums. 🔥
4. 🏆 The driver with the most pole positions is Lewis Hamilton with 107 pole positions. 🏆
5. 🔝 The driver with the most races led is Lewis Hamilton with 190 races led. 🔝
6. 🎯 The driver with the most laps led is Lewis Hamilton with 5487 laps led. 🎯
7. 🏁 Lewis Hamilton and Michael Schumacher both have 7 world titles. Will Hamilton be able to break the record in his new stint with Ferrari? 🏁
8. 🪄 Lewis Hamilton has won in 16 different seasons. 🪄
9. 🥇 Lewis Hamilton has won 9 times at Silverstone Circuit. 🥇
10. 🎖️ Lewis Hamilton and Jacques Villeneuve share the record of winning 4 races in their debut season 🎖️
11. 🚀 Lewis Hamilton achieved 5 pole positions in his debut season. 🚀
12. 🏎️ Most Grand Prix with the same constructor: Lewis Hamilton with 246 races in Mercedes. 🏎️
13. 🏁 The driver with the most consecutive races finished with points is Lewis Hamilton with 48 consecutive races. 🏁
14. 🚀 The driver with the most victories on different circuits is Lewis Hamilton with 31 victories. 🚀
15. 💥 The driver with the most podium finishes at different circuits is Lewis Hamilton with 36 podium finishes. 💥